birdhouse/pavics-compose.sh

```diff
@@ -25,10 +26,17 @@ VARS='
   $POSTGRES_PAVICS_PASSWORD
   $POSTGRES_MAGPIE_USERNAME
   $POSTGRES_MAGPIE_PASSWORD
+  $SERVER_NAME
+  $SERVER_SYNOPSIS
+  $SERVER_INSTITUTION
+  $SUPPORT_EMAIL
+  $SUPPORT_URL
+  $DATA_PERSIST_ROOT
```

Comment:
Well that's the point, I don't want to force it. You can override if you want. If not, leave them as is and default are used.
But if they don't exist at all (as it can be the case when in `OPTIONAL_VARS`), they will break the platform or cause missing key, URL, etc. errors.

Reply:
> But if they don't exist at all (as it can be the case when in `OPTIONAL_VARS`)

So this is exactly why I am confused.  Since the value for those new vars are defined in `default.env` how can they don't exist at all?

Are you trying to defend against git checkout failure due to disk full or hard drive failure? Or user somehow manually editing `default.env` and delete those values knowingly?  If yes, I do not think this is the location to defend against those kind of error since those kind of errors can happen to any code anywhere, not just these variables.

Reply:
For backward compat, please define `DATA_PERSIST_ROOT=/data` in `default.env` and move this `DATA_PERSIST_ROOT` into the `OPTIONAL_VARS` list since it has default value.

Without this backward compat, existing prod deployment that already use `/data` will not start because `DATA_PERSIST_ROOT` is missing in their `env.local` but they do not need to override it since they are already ok using `/data`.

Did the pipeline test worked in this state that `DATA_PERSIST_ROOT` is not in `default.env` and not in `env.local` but is in mandatory `VARS` list?

Reply:
The point is that they must exist, not that they won't possibility exist because of defaults. 

If for some reason `DATA_PERSIST_ROOT` gets renamed or removed at some point, having them in `OPTIONAL_VARS` will silently pass, and then the script will raise some permission error when trying to create dirs. 
Worst, it gets executed as root and pass completely silently and things break without much harder investigations to find the cause. 

Having the value in `VARS` forces `pavics-compose` to ensure they are found and *defined*. Imagine someone incorrectly overrides `env.local` with their custom variable `DATA_PERSIST_ROOT="${SOME_MISSING_VAR}"`, but because of bad order in sourcing files, it resolves as `""`. I think the explicit error is much better and will help anyone quickly identify the causes.

Reply:
> If for some reason `DATA_PERSIST_ROOT` gets renamed or removed at some point, having them in `OPTIONAL_VARS` will silently pass, and then the script will raise some permission error when trying to create dirs.
> Worst, it gets executed as root and pass completely silently and things break without much harder investigations to find the cause.
> 
> Having the value in `VARS` forces `pavics-compose` to ensure they are found and _defined_. Imagine someone incorrectly overrides `env.local` with their custom variable `DATA_PERSIST_ROOT="${SOME_MISSING_VAR}"`, but because of bad order in sourcing files, it resolves as `""`. I think the explicit error is much better and will help anyone quickly identify the causes.

Okay but these problem of accidental rename a var (this must somehow pass PR review) or accidental overriding the default value to empty value in `env.local` can happen to every variables.  Are you suggesting putting all vars from `OPTIONAL_VARS` and all vars in `default.env` into `VARS` as well so they are checked for not empty value?

What if the value is not empty but not valid?

Where do we stop?

I am not against making the code more error resilient and robust.  But it's probably outside the of scope of this PR.  Let's keep the consistency in this PR and we can discuss this issue somewhere else.

Reply:
Not every var, but just the ones that require to be explicitly checked for undefined or empty.
Some variables are valid as empty (eg: not used feature), those should remain in `OPTIONAL_VARS`. 

`DATA_PERSIST_ROOT` is a special use case that cannot be empty, but can use a default if omitted.
So explicit undefined is ok, empty is not. This is what I meant originally that there are 3 states, and only 2 are checked by `VARS` and `OPTIONAL_VARS` (by themselves). Adding the variable in `VAR` + in `default.env` fulfills that 3rd state of undefined OK but not empty via default.

In fact, the real way to fix it *should* be to make it required, but is not for backward compatibility.

Reply:
I hear you.  We can make the code more robust and error resilient but in a different issue to avoid dragging this PR further.

`VARS` and `OPTIONAL_VARS` was for really initially for templating use and we want the list to be as minimal as possible to avoid template expanding a var that was not meant to be.

For your needs, I see maybe a 3rd var `CHECK_NOT_EMPTY` that will contain all vars in `default.env` that should not be empty so we do not hijack `VARS` for the wrong intended use.  Some auditing involved as not every var is invalid if empty.

Reply:
A new list for the 3rd case might not look like a big difference here, but it does impact more when reaching weaver feature. 

I need to dynamically extend the `VARS` here: 
https://github.com/bird-house/birdhouse-deploy/blob/feature/DAC-300-weaver-compose/birdhouse/components/weaver/default.env#L5-L21

The reason I need to work with `VARS`, is that I also use `DATA_PERSIST_ROOT`, which then generates paths to volumes for data-persistence, which them MUST exist. 
This dynamic "hack" of `VARS` felt already unusual enough as is, which is why I initially didn't introduce another list of vars that increases even more the complexity of weaver feature. 

But sure, a subsequent PR after all that could be considered.

Reply:
> I need to dynamically extend the `VARS` here:
> https://github.com/bird-house/birdhouse-deploy/blob/feature/DAC-300-weaver-compose/birdhouse/components/weaver/default.env#L5-L21

Well, the 3rd new var `CHECK_NOT_EMPTY` will have be extensible by components, just like `VARS` because each component will need to inject their own new variables.  Each component will have to be able say here are my new variables that should not be empty (add to `CHECK_NOT_EMPTY`) and a subset of those new variables will be used for template expansion (add to `VARS`).

If you worry how to extend that 3rd new var `CHECK_NOT_EMPTY`, no worries !